From 68cef57e9260265229d194f1309cdffd654c95b6 Mon Sep 17 00:00:00 2001
From: =?UTF-8?q?Sebastian=20Dr=C3=B6ge?= <sebastian.droege@collabora.co.uk>
Date: Sat, 26 Sep 2009 12:56:36 +0200
Subject: [PATCH 7/7] decodebin2: Chains with an exposed endpad are complete too

This allows partial group changes, i.e. demuxer2 in the example below
goes EOS but has a next group and audio2 stays the same.

          /-- >demuxer2---->video
demuxer---             \--->audio1
          \--->audio2
---
 gst/playback/gstdecodebin2.c |   11 ++++++++---
 1 files changed, 8 insertions(+), 3 deletions(-)

diff --git a/gst/playback/gstdecodebin2.c b/gst/playback/gstdecodebin2.c
index 4fe62b4..60dbf57 100644
--- a/gst/playback/gstdecodebin2.c
+++ b/gst/playback/gstdecodebin2.c
@@ -2191,7 +2191,7 @@ out:
  *
  * Returns TRUE if the chain is complete, this means either
  * a) This chain is a dead end, i.e. we have no suitable plugins
- * b) This chain ends in an endpad and this is blocked
+ * b) This chain ends in an endpad and this is blocked or exposed
  *
  * Not MT-safe, always call with decodebin expose lock
  */
@@ -2200,7 +2200,12 @@ gst_decode_chain_is_complete (GstDecodeChain * chain)
 {
   gboolean complete = FALSE;
 
-  if (chain->deadend || (chain->endpad && chain->endpad->blocked)) {
+  if (chain->deadend) {
+    complete = TRUE;
+    goto out;
+  }
+
+  if (chain->endpad && (chain->endpad->blocked || chain->endpad->exposed)) {
     complete = TRUE;
     goto out;
   }
@@ -2543,7 +2548,7 @@ gst_decode_chain_expose (GstDecodeChain * chain, GList ** endpads)
     return TRUE;
 
   if (chain->endpad) {
-    if (!chain->endpad->blocked)
+    if (!chain->endpad->blocked && !chain->endpad->exposed)
       return FALSE;
     *endpads = g_list_prepend (*endpads, gst_object_ref (chain->endpad));
     return TRUE;
-- 
1.6.4.3

